megatron/training.py

```diff
@@ -610,126 +697,166 @@ def setup_model_and_optimizer(model_provider_func,
                 else:
                     train_samples = args.train_iters * args.global_batch_size
+                log.info(f'{train_samples=}')
                 # eval_iters and test_iters here are not actually used, only for
                 # satisfying the input of build_train_valid_test_datasets_provider.
                 # We only need to build the training data here. And we follow
                 # baseline's logic to build eval/test dataset later in
                 # build_train_valid_test_data_iterators.
-                eval_iters = (args.train_iters // args.eval_interval + 1) * \
-                            args.eval_iters
+                eval_iters = (
+                    args.train_iters // args.eval_interval + 1
+                ) * args.eval_iters
                 test_iters = args.eval_iters
-                train_val_test_num_samples = [train_samples,
-                                            eval_iters * args.global_batch_size,
-                                            test_iters * args.global_batch_size]
+                train_val_test_num_samples = [
+                    train_samples,
+                    eval_iters * args.global_batch_size,
+                    test_iters * args.global_batch_size,
+                ]
+                log.info(f"{train_val_test_num_samples=}")
                 # Build the datasets.
                 train_ds, _, _ = build_train_valid_test_datasets_provider(
-                    train_val_test_num_samples)
+                    train_val_test_num_samples
+                )
             with Profile("deepspeed.initialize"):
-                model, optimizer, args.deepspeed_dataloader, opt_param_scheduler = deepspeed.initialize(
+                model, optimizer, args.deepspeed_dataloader, opt_param_scheduler = (
+                    deepspeed.initialize(
+                        model=model[0],
+                        optimizer=optimizer,
+                        args=args,
+                        lr_scheduler=opt_param_scheduler,
+                        training_data=train_ds,
+                        mpu=mpu if args.no_pipeline_parallel else None,
+                        config=args.deepspeed_config_dict,
+                    )
+                )
+            model.set_data_post_process_func(data_post_process)
+        else:
+            log.info(
+                "Did NOT catch: ('args.data_efficiency_curriculum_learning' "
+                "and 'build_train_valid_test_datasets_provider is not None')"
+            )
+            tds0 = time.time()
+            if os.environ.get("PYINSTRUMENT_PROFILER", None):
+                profiler = ez.profile.get_context_manager(
+                    rank=RANK, outdir=args.save
+                )
+            else:
+                profiler = Profile("deepspeed.initialize")
+            log.info("Calling 'deepspeed.initialize'...")
+            log.info(f"Wrapped with: {profiler=}")
+            with profiler:
```

Comment:
Profiling `deepspeed.initialize(...)`

- if running with `PYINSTRUMENT_PROFILER=1`:
  - wrap with [`ezpz.profile.PyInstrumentProfiler`](https://github.com/saforem2/ezpz/blob/c905d0741b0fa9c4d699793cfa500a990494ced6/src/ezpz/profile.py#L65)[^ezpz]

- otherwise:
  - wrap with [`megatron.utils.Profile`](https://github.com/argonne-lcf/Megatron-DeepSpeed/blob/main/megatron/utils.py)

[^ezpz]: From [`ezpz.profile`](https://github.com/saforem2/ezpz/blob/main/src/ezpz/profile.py)